Proposal: build feature to include emoji-db?

black-puppydog opened the pull request:
Hi,

First off: thanks for making this! 🙂 
I am re-implementing a markdown parser for a niche dialect used in a community I'm part of. I'm writing against the test suite from the JS reference implementation, and it turns out that that reference implementation uses a superset of the shortcodes supported here. Actually it just adds the ones from [emoji-data](https://github.com/iamcal/emoji-data/) which include aliases like `:robot_face:`.

So I made a fork of gh-emoji, added the corresponding json file (I found a submodule overkill; it pulls in a whopping 3GB of data 0_o) and modified the generation files.

I'm writing here mostly to see if you'd be interested to develop this into a feature, or if I should just throw it on crates.io myself.

I could spend some time figuring out how to do this cleanly. Off the top of my head, I'd probably try generating two files and then including one or the other in `lib.rs` based on whether the feature is active or not.

But if you think that this is out of scope here, absolutely no offense taken. :)

black-puppydog committed:
Add emoji-data file

black-puppydog committed:
cargo fmt

black-puppydog committed:
Load both emoji datasets

black-puppydog committed:
leave note about opening issues for dependency updates

black-puppydog committed:
Update crate details for publication

black-puppydog committed:
Version the generated data

Not beautiful, but I don't see how one could use this module directly from
git otherwise. I wouldn't want to force anyone developing this as a
dependency (like myself right now) to publish a crate for their modification.

kornelski commented:
I think it'd be best if you fork the crate. I wouldn't want to expand by default beyond GitHub's set, and I don't see a way to include other sets without adding them to the crate ahead of time.

black-puppydog committed:
Extract emojidb generation into function

black-puppydog committed:
Extract genmoji generation into function

black-puppydog committed:
Extract source code generation into separate function

black-puppydog committed:
Keep gemoji, emojidb, and their combination as vectors

Ugly but effective: make copies of all the emojis so we can re-use the
original vectors.

My borrow-checker-foo is just not strong enough for re-use. This could
probably be done much more elegantly...

black-puppydog committed:
Generate & use two versions of the library

black-puppydog commented:
So I went ahead and built this anyway. I still understand that you want to keep the scope manageable, but I thought I'd still illustrate what this would look like. I updated the branch so that there is an `emojidb` feature in the crate, which allows to enable/disable the larger shortcode list at compile time.
The trade-off is that indeed, there are now two large files generated & versioned. But on the other hand, that might enable you to get rid of the genmoji submodule? That's a few 100MB right there...

I'm using this for my little markdown parser over here: https://gitlab.com/dwynen/ssb-markdown-rs/-/blob/main/Cargo.toml

But I should also point out that since emoji-db is used in node-emoji, the `:robot_face:` shortcode does work in e.g. slack, so I might not be the only potential user after all.

Your call. My main branch is ready to publish under a new name, but I could just change the name back and we can integrate this here.

black-puppydog committed:
Mention emoji-db in code comment

black-puppydog committed:
Remove gemoji submodule and replace with shallow copy of file

Cloning literally hundreds of MBs for a single json file just was too
inconvenient for me. Plus, every time the upstream would change there
would still be manual effort necessary to pull the changes in. Just instead
of juggling a submodule, now the effort is to just download a new version
of the file.

black-puppydog committed:
Mention emojidb in the package description

black-puppydog committed:
Update readme to clarify relationship with gh-emoji

black-puppydog commented:
Went ahead and made it a separate crate. I hope I made the relationship with this one here sufficiently clear to avoid any confusions and/or misguided issues filed, let me know if there's anything you'd like to see changed. 🙂
